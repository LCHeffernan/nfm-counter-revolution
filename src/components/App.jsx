import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import "../styles/App.css";
import Content from "./Content";
import DiscController from "./DiscController";

const App = ({ figures }) => {
  const baseAnimationDurationMs = 1200;
  const appIdleTimerDurationMillis = 5000;
  const [index, setIndex] = useState(0);
  const [isAppIdle, setIsAppIdle] = useState(true);
  const [appIdleTimer, setAppIdleTimer] = useState();

  const incrementIndex = () =>
    setIndex((prev) => (prev < figures.length - 1 ? prev + 1 : 0));

  const startAppIdleTimer = () => {
    setIsAppIdle(false);
    setAppIdleTimer(clearTimeout(appIdleTimer));
  };

  useEffect(() => {
    const doc = document.documentElement;
    doc.style.setProperty(
      "--base-animation-duration",
      `${baseAnimationDurationMs}ms`
    );
  }, []);

  useEffect(() => {
    if (!(isAppIdle || appIdleTimer)) {
      setAppIdleTimer(
        setTimeout(() => {
          setIsAppIdle(true);
          setAppIdleTimer(clearTimeout(appIdleTimer));
        }, appIdleTimerDurationMillis)
      );
    }
    return () => clearTimeout(appIdleTimer);
  }, [isAppIdle, appIdleTimer]);

  return (
    <div className="App">
      <main>
        <DiscController
          incrementIndex={incrementIndex}
          setIndex={setIndex}
          index={index}
          figures={figures}
          baseAnimationDurationMs={baseAnimationDurationMs}
          startAppIdleTimer={startAppIdleTimer}
          isAppIdle={isAppIdle}
        />
        <Content
          figures={figures}
          index={index}
          baseAnimationDurationMs={baseAnimationDurationMs}
          isAppIdle={isAppIdle}
        />
      </main>
    </div>
  );
};

App.propTypes = {
  figures: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      angle: PropTypes.number,
      text: PropTypes.arrayOf(PropTypes.string),
    })
  ).isRequired,
};

export default App;
