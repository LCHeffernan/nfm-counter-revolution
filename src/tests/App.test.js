import React from "react";
import { render, screen } from "@testing-library/react";
import App from "../components/App";

describe("App", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("is rendered", () => {
    render(<App />);

    expect(screen.getByTestId("app")).toBeInTheDocument();
  });
});
