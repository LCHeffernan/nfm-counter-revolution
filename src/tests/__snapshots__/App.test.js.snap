// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Ap matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
    data-testid="app"
  >
    <h1>
      Counter Revolution
    </h1>
  </div>
</DocumentFragment>
`;

exports[`App App matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
    data-testid="app"
  >
    <h1>
      Counter Revolution
    </h1>
  </div>
</DocumentFragment>
`;

exports[`App matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
    data-testid="app"
  >
    <h1>
      Counter Revolution
    </h1>
  </div>
</DocumentFragment>
`;
